### PR TITLE
feat(replies): enable based on redaction implementation

### LINKED ISSUE #86: Feature Request: Message Replies
Just like the reply feature on Riot, a way to reference a particular message in your message. It could function similar to the /redact command, where you type /reply, then tab through the messages until you see the message text for the message you wish to reply to. Then, whatever you type after that would be the reply. 

Example:
`/reply $38j9328jd98j983jd923:"What day is Memorial day?" May 27th`

The resulting message should look just like how a reply would look when done on Riot.

### PR DESCRIPTION
**Pull request by myii**
* Close #86
* https://matrix.org/docs/spec/client_server/latest#rich-replies

---

@poljar As mentioned in https://github.com/poljar/weechat-matrix/issues/86#issuecomment-626198269, I'm fully expecting there to be changes required here, so feel free to point out any issues.

### CODE REVIEW

File: matrix/commands.py [hunk 1]
```diff
@@ -1321,6 +1342,64 @@ def predicate(event_id, line):
...
+        if event_tag in tags:
+            return True
+        return False
+
+    for server in SERVERS.values():
+        if buffer in server.buffers.values():
+            room_buffer = server.find_room_from_ptr(buffer)
+
+            # event_id, reply = parse_reply_args(args)
+            event_id, reply = parse_redact_args(args)
```
Comment (myii) [inline matrix/commands.py:1360]: This is going to have to change, I've just taken advantage of `parse_redact_args` for now, perhaps erroneously.

**Comment by balsoft** [comment]
Can you give some usage instructions?

Currently building this to try it.

**Comment by balsoft** [comment]
Hrm, I get
```
23:44 =!= Error: another command "reply" already exists for plugin "python"
```

Because I already have `wee-slack` installed. What do you think is the best solution here?

**Comment by myii** [comment]
> Can you give some usage instructions?

@balsoft Usage is the same as `redact`, you type `/retry` and then keep hitting `TAB` until it reaches the message that you want to reply to and then enter the reply straight after that.  For example:

```
/reply $1588763480399Vmcyh:matrix.org|"Preview of the message to reply to.." This is the beginning of my reply...
```

> ```
> 23:44 =!= Error: another command "reply" already exists for plugin "python"
> ```
> 
> Because I already have `wee-slack` installed. What do you think is the best solution here?

I'm also using this alongside `wee-slack` and I haven't seen that so far (not saying it doesn't happen, though).

**Comment by balsoft** [comment]
@myii I can't seem to get it working with wee-slack.

**Comment by myii** [comment]
@balsoft OK, this is a simple thing to try.  Just change this line to anything else:

* https://github.com/poljar/weechat-matrix/pull/184/files#diff-51ccfb82ea67ee80dedfb814eeb8f585R290

So let's say you change it to `"rep",`, then you will type `/rep` followed by a space and then hit TAB until the right message comes up.

You need to ensure you're on a Matrix buffer as well (obviously).

**Comment by balsoft** [comment]
Hmm, now it just doesn't do anything (it sends the message, but there isn't a "reply" attached to it)

**Comment by poljar** [comment]
> Hrm, I get
> 
> ```
> 23:44 =!= Error: another command "reply" already exists for plugin "python"
> ```
> 
> Because I already have `wee-slack` installed. What do you think is the best solution here?

This is an ongoing problem we have been facing because this is a Weechat Python script vs a Weechat plugin. Stuff we define is namespaced under the Weechat Python plugin, so if two scripts define the same command one of them won't work.

Perhaps we should name the command `/reply-matrix` as a workaround. A rewrite of this to become a Weechat plugin so those problems will go away in the plugin.

**Comment by myii** [comment]
@balsoft @poljar Here's a GIF showing how it's working at my end, using `/reply` even with `wee-slack` active:

Peek 2020-05-10 11-19 OUT

* Note, when I use `/reply` in a Slack channel, I get the options provided by `wee-slack`.

---

Further details that might help get to the bottom of this:

```bash
./aesthetic.py:      SCRIPT_VERSION = '1.0.5'
./colorize_nicks.py: SCRIPT_VERSION = "26"
./go.py:             SCRIPT_VERSION = '2.6'
./grep.py:           SCRIPT_VERSION = "0.8.2"
./matrix.py: WEECHAT_SCRIPT_VERSION = "0.2.0"
./typing_counter.py: SCRIPT_VERSION = "1.0"
./url_hint.py:       SCRIPT_VERSION = "0.8"
./urlbar.py:         SCRIPT_VERSION = "13"
./vimode.py:         SCRIPT_VERSION = "0.7"
./wee_slack.py:      SCRIPT_VERSION = "2.5.0"
```

```bash
Package            Version
------------------ ----------
aiofiles           0.5.0
aiohttp            3.6.2
async-timeout      3.0.1
atomicwrites       1.4.0
attrs              19.3.0
cachetools         4.1.0
certifi            2020.4.5.1
cffi               1.14.0
chardet            3.0.4
cryptography       2.9.2
future             0.18.2
h11                0.9.0
h2                 3.2.0
hpack              3.0.0
hyperframe         5.2.0
idna               2.9
importlib-metadata 1.6.0
jsonschema         3.2.0
Logbook            1.5.3
matrix-nio         0.10.0
multidict          4.7.5
peewee             3.13.3
pip                20.1
pkg-resources      0.0.0
pycparser          2.20
pycryptodome       3.9.7
Pygments           2.6.1
pyOpenSSL          19.1.0
pyrsistent         0.16.0
python-magic       0.4.18
python-olm         3.1.3
requests           2.23.0
setuptools         46.1.3
six                1.14.0
unpaddedbase64     1.1.0
urllib3            1.25.9
webcolors          1.11.1
wheel              0.34.2
yarl               1.4.2
zipp               3.1.0
```

---

@poljar What would you like me to do with this PR in the meantime?

**Comment by poljar** [comment]
Just renaming the command should get rid of the clash, not sure why you're not experiencing it.

**Comment by myii** [comment]
> Just renaming the command should get rid of the clash, not sure why you're not experiencing it.

@poljar Renamed to `reply-matrix` as suggested and force-pushed here.  It makes it a bit slower to start the reply but as long as it works.  Anything else that needs doing here?

File: matrix/commands.py [hunk 2]
```diff
@@ -1321,6 +1342,64 @@ def predicate(event_id, line):
...
+                    "No such message with event id "
+                    "{event_id} found.".format(event_id=event_id))
+                return W.WEECHAT_RC_OK
+
+            formatted_data = Formatted.from_input_line(reply)
+            server.room_send_message(
+                room_buffer,
+                formatted_data,
+                "m.text",
+                ignore_unverified_devices=True,
```
Comment (poljar) [inline matrix/commands.py:1385]: I don't think we should do that considering the behavior of the script in other places.

**Comment by balsoft** [comment]
Huh, they kinda work, but don't show in RiotX for some reason...

File: matrix/commands.py [hunk 3]
```diff
@@ -1321,6 +1342,64 @@ def predicate(event_id, line):
...
+
+        if event_tag in tags:
+            return True
+        return False
+
+    for server in SERVERS.values():
+        if buffer in server.buffers.values():
+            room_buffer = server.find_room_from_ptr(buffer)
+
+            # event_id, reply = parse_reply_args(args)
```
Comment (poljar) [inline matrix/commands.py:1359]: This should be gone.

**Comment by poljar** [comment]
They don't work in encrypted rooms yet, but a nio release is happening today which will fix that.

File: matrix/commands.py (hunk 2 above)
Comment (myii) [reply matrix/commands.py]: OK, line removed.

File: matrix/commands.py (hunk 3 above)
Comment (myii) [reply matrix/commands.py]: Removed, replaced with a comment instead explaining that use of `parse_redact_args` isn't an oversight.

**Comment by myii** [comment]
@poljar Made the requested changes.  I also noticed you have a couple of items mentioned in the Travis build config validation:

* https://travis-ci.org/github/poljar/weechat-matrix/builds/685282618/config

Having dealt with that in the org I maintain, I've got another commit for it if you want it, which clears those issues:

* https://travis-ci.com/github/myii/weechat-matrix/builds/165006506/config

**Comment by poljar** [comment]
> @poljar Made the requested changes. I also noticed you have a couple of items mentioned in the Travis build config validation:
> 
>     * https://travis-ci.org/github/poljar/weechat-matrix/builds/685282618/config
> 
> 
> Having dealt with that in the org I maintain, I've got another commit for it if you want it, which clears those issues:
> 
>     * https://travis-ci.com/github/myii/weechat-matrix/builds/165006506/config

Sure.

**Comment by myii** [comment]
@poljar Done, added in 62507ac26694df340a13fd96937f807c62dbe647.

**Comment by myii** [comment]
> Huh, they kinda work, but don't show in RiotX for some reason... Interesting

@balsoft OK, I've looked at the actual events itself and I can see that the current implementation isn't entering this `if` block to get the `format` and `formatted_body` applied:

https://github.com/poljar/weechat-matrix/blob/69ad2a9c03d516c212d3d0700dbb2bfe654f6365/matrix/server.py#L1129-L1131

Which is checked here:

https://github.com/poljar/weechat-matrix/blob/a1abfdc3d625221f065df5a1048b5a824fd43eaa/matrix/colors.py#L75-L80

Which links back to this:

https://github.com/poljar/weechat-matrix/blob/a1abfdc3d625221f065df5a1048b5a824fd43eaa/matrix/colors.py#L479-L489

So addressing this should fix the problem encountered in RiotX.

**Comment by myii** [comment]
Actually, ignore the previous comment, it's sufficient to modify the `if` itself:

https://github.com/poljar/weechat-matrix/blob/69ad2a9c03d516c212d3d0700dbb2bfe654f6365/matrix/server.py#L1129

To:

```python
        if formatted.is_formatted() or in_reply_to_event_id:
```

And then work with it from there.

What I've found is that some clients are mature enough to take the `event_id` and prepare the reply preview themselves (e.g. riot.im and even Telegram via. `mautrix-telegram` (Telematrix)).  Other clients need all the data passed through to them (e.g. RiotX, FluffyChat).  This isn't exactly _hard_ but it's more work to extract the data from the event being replied to and then preparing it in the necessary formats (for both `content["formatted_body"]` **and** `content["body"]`).  Not to mention then finding the right way to add the preview to `weechat` itself.

**Comment by poljar** [comment]
That's called the reply fallback and is a source of controversy, some people argue that it shouldn't exist. If someone wants to add fallbacks I'm fine with it but this can be merged without them.

**Comment by myii** [comment]
> That's called the reply fallback and is a source of controversy, some people argue that it shouldn't exist. If someone wants to add fallbacks I'm fine with it but this can be merged without them.

@poljar I'm glad you said this.  In that case, I'll leave this to merged as-is and then the reply fallbacks can be tackled another time, if desired.

**Comment by wavexx** [comment]
> @balsoft Usage is the same as `redact`, you type `/retry` and then keep hitting `TAB` until it reaches the message that you want to reply to and then enter the reply straight after that (as mentioned in the [original issue](https://github.com/poljar/weechat-matrix/issues/86#issue-446745457)). For example:

Tangential, but I wonder if there is a way to make it work better if you have a global default of ``weechat.completion.partial_completion_command_arg=off`` (I generally prefer to see the list of completions, but in this case it's really broken without)

**Comment by myii** [comment]
@poljar I've done a final force push with this diff for the main commit:

```diff
------------------------------- matrix/server.py -------------------------------
index a6ed60d..7d9ee0f 100644
@@ -1129,10 +1129,9 @@ class MatrixServer(object):
 
-        if formatted.is_formatted():
+        if formatted.is_formatted() or in_reply_to_event_id:
             content["format"] = "org.matrix.custom.html"
             content["formatted_body"] = formatted.to_html()
-
-        if in_reply_to_event_id:
-            content["m.relates_to"] = {
-                "m.in_reply_to": {"event_id": in_reply_to_event_id}
-            }
+            if in_reply_to_event_id:
+                content["m.relates_to"] = {
+                    "m.in_reply_to": {"event_id": in_reply_to_event_id}
+                }
```

This ensures that replies are sending through `content["format"]` and `content["formatted_body"]` as mentioned above.  It also allows for sharing some follow-up ideas with @balsoft, to get at least some feedback in RiotX (and even FluffyChat).

**Comment by myii** [comment]
> Tangential, but I wonder if there is a way to make it work better if you have a global default of `weechat.completion.partial_completion_command_arg=off` (I generally prefer to see the list of completions, but in this case it's really broken without)

@wavexx Note, this PR is now using `/reply-matrix [TAB]`, as mentioned in https://github.com/poljar/weechat-matrix/pull/184#issuecomment-626312035.

**Comment by myii** [comment]
Thanks for the merge.

---

@balsoft I was supposed to share this earlier but perhaps it's even easier now that this has been merged.  I did some testing with RiotX and this is a very basic starting point that you might be interested in trying out:

```diff
diff --git a/matrix/server.py b/matrix/server.py
index ac20e3e..b7a25b1 100644
--- a/matrix/server.py
+++ b/matrix/server.py
@@ -1134,6 +1134,16 @@ class MatrixServer(object):
                 content["m.relates_to"] = {
                     "m.in_reply_to": {"event_id": in_reply_to_event_id}
                 }
+                content["formatted_body"] = (
+                    "<mx-reply><blockquote>"
+                    "In reply to <a href=\"https://matrix.to/#/{}/{}\">this message</a>"
+                    "</blockquote></mx-reply>{}".format(
+                        room.room_id,
+                        in_reply_to_event_id,
+                        content["formatted_body"],
+                    )
+                )
+                content["body"] = Formatted.from_html(content["formatted_body"]).to_plain()
 
         try:
             uuid = self.room_send_event(
```

All this does for now is indicate that the message is reply and shows a link that sends you back to the original message.  To get the user and original message, we'd have to pull in the details of the original message using the event ID (unless there's already something available here).

The structure itself is a replication of the data that shows up when looking at the `source` in riot.im.  Perhaps there's an easier way of grabbing that as well.

**Comment by 2Belette** [comment]
Just wondering what is the status for this one ? I just tested and current implementation is still not showing properly on latest version of Element mobile

**Pull request merged**